lms/djangoapps/courseware/management/commands/fix_student_module_newlines.py

```diff
@@ -0,0 +1,330 @@
+
+        row_to_keep = self.row_to_keep(correct_course_row, newline_course_row)
+
+        # To minimize the chances of conflicts and preserve history, if we
+        # decided to keep the data from the newline_course_row, we copy that
+        # data into correct_course_row, and then delete the newline_course_row
+        # as a cleanup step.
+        if row_to_keep.id == newline_course_row.id:
+            log.info(
+                "Conflict: Choosing data from newline course, copying data from CSM %s to %s - (%s) %s",
+                newline_course_row.id,
+                correct_course_row.id,
+                newline_course_row.module_type,
+                newline_course_row.module_state_key
+            )
+            correct_course_row.state = newline_course_row.state
+            correct_course_row.grade = newline_course_row.grade
+            correct_course_row.max_grade = newline_course_row.max_grade
+
+            if dry_run:
+                return FixResult(record_trimmed=0, data_copied=1, record_deleted=1, capa_merge=0, error=0)
+
+            try:
+                with transaction.atomic():
+                    correct_course_row.save()
+                    self.delete_row(newline_course_row)
+                return FixResult(record_trimmed=0, data_copied=1, record_deleted=1, capa_merge=0, error=0)
+            except DatabaseError:
+                log.exception(
+                    "Failed while trying save CSM row %s and delete row %s",
+                    correct_course_row.id,
+                    newline_course_row.id
+                )
+                return FixResult(record_trimmed=0, data_copied=0, record_deleted=0, capa_merge=0, error=1)
+
+        # If we've reached this point, we just want to keep correct_course_row
+        # and delete the newline_course_row entry.
+        log.info(
+            "Conflict: Choosing data from record with correct course_id (%s) "
+            "and deleting row with newline in course_id (%s) - (%s) %s",
+            correct_course_row.id,
+            newline_course_row.id,
+            newline_course_row.module_type,
+            newline_course_row.module_state_key
+        )
+
+        if dry_run:
+            return FixResult(record_trimmed=0, data_copied=0, record_deleted=1, capa_merge=0, error=0)
+
+        try:
+            with transaction.atomic():
+                self.delete_row(newline_course_row)
+            return FixResult(record_trimmed=0, data_copied=0, record_deleted=1, capa_merge=0, error=0)
+        except DatabaseError:
+            log.exception("Could not delete CSM row %s", newline_course_row.id)
+            return FixResult(record_trimmed=0, data_copied=0, record_deleted=0, capa_merge=0, error=1)
+
+    def delete_row(self, model):
+        """Delete the row and history references to it."""
+        for history_row in BaseStudentModuleHistory.get_history([model]):
+            history_row.delete()
```

Comment:
I thought that removal was the consensus. I have to kill the history rows in order to kill the row in CSM because of foreign key constraints in local testing (I realize it doesn't apply when we run it on prod because we're using the extended history table there). It's true that we could rename the course keys in the CSM row instead.